pkg/sql/alter_job_owner.go

```diff
@@ -0,0 +1,168 @@
+	if err != nil {
+		return nil, err
+	}
+	return &alterJobOwnerNode{
+		jobID: typedJobID,
+		owner: owner,
+	}, nil
+}
+
+func (n *alterJobOwnerNode) startExec(params runParams) error {
+	ctx := params.ctx
+	p := params.p
+	newOwner := n.owner
+
+	// The top-level owner column this updates was added in 25.1.
+	v, err := p.InternalSQLTxn().GetSystemSchemaVersion(ctx)
+	if err != nil {
+		return err
+	}
+	if !v.AtLeast(clusterversion.V25_1.Version()) {
+		return pgerror.Newf(pgcode.FeatureNotSupported, "ALTER JOB OWNER requires version %s", clusterversion.V25_1)
+	}
+
+	exprEval := p.ExprEvaluator("ALTER JOB OWNER")
+	jobIDInt, err := exprEval.Int(ctx, n.jobID)
+	if err != nil {
+		return err
+	}
+	jobID := jobspb.JobID(jobIDInt)
+
+	cur, err := params.ExecCfg().InternalDB.Executor().QueryRowEx(
+		ctx, "get-job-owner", p.txn, sessiondata.NodeUserSessionDataOverride,
+		`SELECT owner FROM system.jobs WHERE id = $1`, jobID,
+	)
+	if err != nil {
+		return err
+	}
+	if cur == nil {
+		return jobs.NewJobNotFoundError(jobID)
+	}
+
+	currentOwner := username.MakeSQLUsernameFromPreNormalizedString(string(tree.MustBeDString(cur[0])))
+
+	if err := n.checkOwnership(ctx, params.p, currentOwner, false); err != nil {
+		return err
+	}
+
+	if err := n.checkOwnership(ctx, params.p, newOwner, true); err != nil {
+		return err
+	}
+
+	if newOwner == currentOwner {
+		return nil
+	}
+
+	if _, err = params.ExecCfg().InternalDB.Executor().ExecEx(
+		ctx, "set-job-owner", p.txn, sessiondata.NodeUserSessionDataOverride,
+		`UPDATE system.jobs SET owner = $2 WHERE id = $1`, jobID, newOwner.Normalized(),
+	); err != nil {
+		return err
+	}
+
+	// Update the legacy payload's copy as well since there are still some paths
+	// which read it.
+	// TODO(dt): remove this when we drop the field from the legacy payload.
+	legacyJob, err := params.ExecCfg().JobRegistry.LoadJobWithTxn(ctx, jobID, p.InternalSQLTxn())
+	if err != nil {
+		return err
+	}
+	return legacyJob.WithTxn(p.InternalSQLTxn()).Update(ctx, func(_ isql.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
+		md.Payload.UsernameProto = newOwner.EncodeProto()
+		ju.UpdatePayload(md.Payload)
+		return nil
+	})
+}
+
+// checkOwnership checks that the current user is allowed to set the owner of a
+// job that is owned or will be owned by the passed owner. This should be called
+// twice: once for the current owner and once for the new owner, with the param
+// forNewOwner indicating which is being done to provide clearer errors.
+func (n *alterJobOwnerNode) checkOwnership(
+	ctx context.Context, p *planner, owner username.SQLUsername, forNewOwner bool,
+) error {
+	// If the user is owner, this is trivially true.
+	if owner == p.User() {
+		return nil
+	}
+
+	// Nobody is allowed to transfer jobs to "node".
+	if owner == username.NodeUserName() {
+		return pgerror.Newf(pgcode.InsufficientPrivilege, "cannot transfer ownership to or from node")
```

Comment:
I don't think typical cockroach users know about the node user. Perhaps rephrase to "cannot transfer ownership to or from the 'node' user, used by cockroach internally"

Reply:
You cannot create, or use, or even connect as the `node` user already, so I don't think you'd be trying to transfer anything to or from it anyway unless you'd already decided to venture into internals so I think by the point you get this error, you know you're messing with internals already.

Reply:
Made it "internal node user"